implementations/micrometer-registry-dynatrace/src/main/java/io/micrometer/dynatrace/v2/DynatraceExporterV2.java

```diff
@@ -0,0 +1,383 @@
+                    }
+                    return null;
+                })
+                .filter(Objects::nonNull);
+    }
+
+    private Metric.Builder createMetricBuilder(Meter meter) {
+        return metricBuilderFactory
+                .newMetricBuilder(meter.getId().getName())
+                .setDimensions(fromTags(meter.getId().getTags()))
+                .setTimestamp(Instant.ofEpochMilli(clock.wallTime()));
+    }
+
+    private static DimensionList fromTags(List<Tag> tags) {
+        return DimensionList.fromCollection(
+                tags.stream()
+                        .map(tag -> Dimension.create(tag.getKey(), tag.getValue()))
+                        .collect(Collectors.toList()));
+    }
+
+    static <T> Stream<T> streamOf(Iterable<T> iterable) {
+        return StreamSupport.stream(iterable.spliterator(), false);
+    }
+
+    private void send(List<String> metricLines) {
+        try {
+            String body = String.join("\n", metricLines);
+            if (logger.isDebugEnabled()) {
+                logger.debug("sending lines:\n" + body);
+            }
+
+            httpClient.post(endpoint)
+                    .withHeader("Authorization", "Api-Token " + config.apiToken())
+                    .withHeader("Content-Type", "text/plain")
+                    .withHeader("User-Agent", "micrometer")
+                    .withPlainText(body)
+                    .send()
+                    .onSuccess((r) -> handleSuccess(metricLines.size(), r))
+                    .onError((r) -> logger.error("Failed metric ingestion. code={} response.body={}", r.code(), r.body()));
+        } catch (Throwable throwable) {
+            logger.error("Failed metric ingestion: {}", throwable.getMessage());
+        }
+    }
+
+    private void handleSuccess(int totalSent, HttpSender.Response r) {
+        if (r.code() == 202) {
+            if (IS_NULL_ERROR_RESPONSE.matcher(r.body()).find()) {
+                Matcher linesOkMatchResult = EXTRACT_LINES_OK.matcher(r.body());
+                Matcher linesInvalidMatchResult = EXTRACT_LINES_INVALID.matcher(r.body());
+                if (linesOkMatchResult.find() && linesInvalidMatchResult.find()) {
+                    logger.info("Sent {} metric lines, linesOk: {}, linesInvalid: {}.",
+                            totalSent, linesOkMatchResult.group(1), linesInvalidMatchResult.group(1));
+                } else {
+                    logger.warn("could not parse response: {}", r.body());
+                }
+            } else {
+                logger.warn("could not parse response: {}", r.body());
+            }
+        } else {
+            // common pitfall if URI is supplied in v1 format (without endpoint path)
+            logger.error("Expected status code 202, got {}. Did you specify the ingest path (e. g. /api/v2/metrics/ingest)?", r.code());
```

Comment:
Should be `e.g.:`
```suggestion
            logger.error("Expected status code 202, got {}. Did you specify the ingest path (e.g.: /api/v2/metrics/ingest)?", r.code());
```

Reply:
As far as I understand, this happens if `error` is not null in the response, right?
I'm wondering if logging `error` out would give more info to the users.

Reply:
This happens if the HTTP status code is not 202, which is the signal for successfully ingested metrics. This can happen if the passed URL is a valid URL, but not a valid ingestion endpoint.

Reply:
If the success code is not a 202 there is no error message. In that case, the response body will contain a whole webpage, not just a json object. Printing that is probably not helpful to the user, and it will clutter the logs.